### PR TITLE
d4xx: Locking on open dfu device

### PR DESCRIPTION
**Pull request by dmipx**
Fix bus clock restore on dfu open/close.

sometimes we have race condition while running tests
```
[109371.797365] d4xx 30-001a: ds5_dfu_device_release(): i2c-2 bus_clk 400000, restore to 100000
[109371.807802] d4xx 30-001a: ds5_dfu_device_open(): i2c-2 bus_clk = 100000, set 400000
[109371.814781] d4xx 30-001a: ds5_dfu_device_release(): i2c-2 bus_clk 400000, restore to 100000
[109371.823742] d4xx 30-001a: ds5_dfu_device_open(): i2c-2 bus_clk = 100000, set 400000
[109371.830182] d4xx 30-001a: ds5_dfu_device_open(): i2c-2 bus_clk = 400000, set 400000
[109371.838771] d4xx 30-001a: ds5_dfu_device_release(): i2c-2 bus_clk 400000, restore to 400000
[109371.847535] d4xx 30-001a: ds5_dfu_device_release(): i2c-2 bus_clk 400000, restore to 400000
[109371.855556] d4xx 30-001a: ds5_dfu_device_open(): i2c-2 bus_clk = 400000, set 400000
[109371.862556] d4xx 30-001a: ds5_dfu_device_release(): i2c-2 bus_clk 400000, restore to 400000
```

### CODE REVIEW

File: kernel/realsense/d4xx.c [hunk 1]
```diff
@@ -4967,26 +4967,29 @@ static int ds5_dfu_device_open(struct inode *inode, struct file *file)
 			state->client->adapter);
 #endif
 #endif
+	mutex_lock(&state->lock);
 	if (state->dfu_dev.device_open_count)
 		return -EBUSY;
```
Comment (Nir-Az) [inline kernel/realsense/d4xx.c:4972]: We need `mutex_unlock(&state->lock);` here right?

File: kernel/realsense/d4xx.c [hunk 2]
```diff
@@ -4997,6 +5000,7 @@ static int ds5_dfu_device_open(struct inode *inode, struct file *file)
 	i2c_set_adapter_bus_clk_rate(parent, DFU_I2C_BUS_CLK_RATE);
 #endif
 #endif
+	mutex_unlock(&state->lock);
```
Comment (Nir-Az) [inline kernel/realsense/d4xx.c:5003]: Who reads `state->dfu_dev.bus_clk_rate`?

We set using the i2c API right?
Why do we need to keep the last value? can't we hard coded set 400 on entrance and set 100 on exit?

**Comment by Nir-Az** [comment]
Nice catch Dima.
Can you write how did you loop enter and exit to DFU mode?
Perhaps this can turn into a unit test for this fix

**Comment by Nir-Az** [comment]
@mengyui On your PR you also added lock and said it caused some crash.
Can you specify on which use case so we can double check here?

**Comment by Nir-Az** [comment]
Another interesting question is how do we see 2 opens w/o a close in between.
How can this happen? Was it teated on a single camera?

**Comment by mengyui** [comment]
The issue which i'm trying to fix is related to "device_open_count"
when try to update FW on the system with multi-9296/D457, the issue may be found.

BTW, should the lock be unlocked before returning from the function?
https://github.com/dmipx/realsense_mipi_platform_driver/blame/busclk-fix/kernel/realsense/d4xx.c#L4972

File: kernel/realsense/d4xx.c (hunk 2 above)
Comment (dmipx) [reply kernel/realsense/d4xx.c:5003]: we can hard code values

**Comment by dmipx** [comment]
> Another interesting question is how do we see 2 opens w/o a close in between. How can this happen? Was it teated on a single camera?

It was tested with unit-tests on single camera.

**Comment by dmipx** [comment]
> The issue which i'm trying to fix is related to "device_open_count" when try to update FW on the system with multi-9296/D457, the issue may be found.
> 
> BTW, should the lock be unlocked before returning from the function? https://github.com/dmipx/realsense_mipi_platform_driver/blame/busclk-fix/kernel/realsense/d4xx.c#L4972

Thanks.
Fixed.

**Comment by dmipx** [comment]
> Nice catch Dima. Can you write how did you loop enter and exit to DFU mode? Perhaps this can turn into a unit test for this fix

This is not "enters" dfu mode, it's open dfu device and close dfu device. For Jetson we switch I2C bus speed clock.
The test can be as follows (for jp4/5 only):
1. verify i2c bus speed 100k
2. open dfu device file
3. verify i2c bus speed 400k
4. close dfu device file
5. verify i2c bus speed 100k

File: kernel/realsense/d4xx.c (hunk 1 above)
Comment (dmipx) [reply kernel/realsense/d4xx.c:4973]: right, this was another adjustment of where mutex should be hold and we missed that.

**Review by Arun-Prasad-V — APPROVED** [review]
LGTM.

**Pull request merged**